### PR TITLE
fix: zip cross directory attack vulnerability

### PR DESCRIPTION
**Pull request by msslulu**
English | [简体中文](https://github.com/opentiny/tiny-engine-backend-java/blob/main/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine-backend-java/blob/main/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened ZIP extraction safety to block path traversal and ensure extracted files remain inside intended directories.
  * Safer directory/file creation and more robust null/empty input handling (e.g., duplicate-removal returns an empty list for null input).

* **Tests**
  * Large expansion of automated tests covering ZIP handling (including traversal protection), data flattening, JSON parsing, file validation, Base64 encode/decode, and many edge cases for improved reliability.

[Review Change Stack](https://app.coderabbit.ai/change-stack/opentiny/tiny-engine-backend-java/pull/312?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: base/src/main/java/com/tinyengine/it/common/utils/Utils.java [hunk 1]
```diff
@@ -222,13 +222,34 @@ private static File convertMultipartFileToFile(MultipartFile multipartFile) thro
     private static List<FileInfo> processZipEntries(ZipInputStream zis, File tempDir) throws IOException {
         List<FileInfo> fileInfoList = new ArrayList<>();
         ZipEntry zipEntry;
-
+        // 将 tempDir 转为规范路径（例如解析符号链接、父目录等）
```
Comment (hexqi) [inline base/src/main/java/com/tinyengine/it/common/utils/Utils.java:228]: 补上测试用例

File: base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java [hunk 2]
```diff
@@ -79,5 +124,883 @@ void testReadFileContent() {
...
+    }
+    @Test
+    public void test_flat_doesNotFlattenListsOrArrays() {
+        Map<String, Object> input = new HashMap<>();
+        input.put("list", Arrays.asList("one", "two"));
+        Map<String, Object> output = Utils.flat(input);
+        assertEquals(2, output.size());
+        assertEquals("one", output.get("list[0]"));
+        assertEquals("two", output.get("list[1]"));
+    }
```
[the remark this answers is not in the document]
Comment (chilingling) [reply base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java:457]: Utils.flat 只会展平嵌套的 Map 值；它不会生成 list[0] / list[1] 这样的键。因此，即使你的代码实现是正确的，这些断言也必然会失败。应该将断言修改为：期望输出的 Map 仍包含单个 "list" 键，且其值等于原始传入的列表/数组。

File: base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java [hunk 3]
```diff
@@ -79,5 +124,883 @@ void testReadFileContent() {
...
+    public void test_encodeObjectToBase64_WithMap() throws Exception {
+        Map<String, Object> map = new HashMap<>();
+        map.put("name", "Alice");
+        map.put("age", 20);
+        String base64 = Utils.encodeObjectToBase64(map);
+        assertNotNull(base64);
+        assertTrue(base64.contains("Alice"));
+        assertTrue(base64.contains("20"));
+        assertTrue(base64.contains("image/"));
+    }
```
[the remark this answers is not in the document]
Comment (chilingling) [reply base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java]: 对于编码后的字符串来说，像 contains("Alice") / contains("image/") 这样的断言是无效的，并会导致测试误报失败。请更新测试用例：先将 Utils.encodeObjectToBase64 编码后的结果解码回原始字节/字符串，然后再去断言解码后的数据中是否包含预期内容。

File: base/src/main/java/com/tinyengine/it/common/utils/Utils.java [hunk 4]
```diff
@@ -219,16 +222,37 @@ private static File convertMultipartFileToFile(MultipartFile multipartFile) thro
...
+            // 拼接并规范化路径
+            Path targetPath = safeDir.resolve(entryName).normalize();
+
+            log.info("Processing ZIP entry: {}, target path: {}", entryName, targetPath);
+
+            // 关键校验：确保目标路径仍在 safeDir 之下
+            if (!targetPath.startsWith(safeDir)) {
+                throw new SecurityException("检测到跨目录攻击: " + entryName);
+            }
             File newFile = new File(tempDir, zipEntry.getName());
```
Comment (chilingling) [inline base/src/main/java/com/tinyengine/it/common/utils/Utils.java]: 路径验证是在 targetPath 上进行的，但文件 I/O 仍然使用 new File(tempDir, zipEntry.getName())。请将所有的文件系统操作都保持在已经验证过的路径上，以避免验证与实际使用之间出现偏差。建议使用 targetPath (或通过 targetPath.toFile() 转换) 替代 newFile 来执行 Files.createDirectories、extractFile 和 readFileContent 操作。

File: base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java [hunk 5]
```diff
@@ -13,25 +13,70 @@
...
+    // 假设 validateFileStream 是某个验证类的静态方法，这里用模拟
+    private MockedStatic<Utils> validationUtilMock;
+    @BeforeEach
+    void setUp() {
+        fileUtilMock = mockStatic(FileUtil.class);
+        filesMock = mockStatic(Files.class);
+        jsonUtilsMock = mockStatic(JsonUtils.class);
+        mimeTypeMock = mockStatic(Enums.MimeType.class);
+        validationUtilMock = mockStatic(Utils.class);
+    }
```
[the remark this answers is not in the document]
Comment (chilingling) [reply base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java]: 修复 UtilsTest 中的 Utils 静态 Mock 问题，需使用 validationUtilMock.when(() -> Utils.*) 语法进行 Stubbing(插桩)。
测试代码使用 Mockito 的常规语法来 stub 静态方法，这些调用并没有注册到当前激活的 MockedStatic 上。因此，在真实调用 Utils.unzip() 时，这些 Mock stub将无法可靠地生效。

File: base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java [hunk 6]
```diff
@@ -79,5 +124,883 @@ void testReadFileContent() {
...
+        assertThat(new String(result, StandardCharsets.UTF_8)).isEqualTo("test content");
+    }
+    @Test
+    public void test_readAllBytes_Normal() throws IOException {
+        byte[] expectedBytes = "test data".getBytes();
+        InputStream mockInputStream = Mockito.mock(InputStream.class);
+        when(mockInputStream.read(any(byte[].class))).thenReturn(5, -1);
+        byte[] result = Utils.readAllBytes(mockInputStream);
+        assertArrayEquals(expectedBytes, result);
+    }
```
[the remark this answers is not in the document]
Comment (chilingling) [reply base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java:916]: readAllBytes(InputStream) 会从你提供的 byte[] 缓冲区中读取并写入数据。在测试中，Mock 对象只返回了长度（例如 thenReturn(5, -1)），但从未将期望的内容复制进缓冲区中，导致产生的字节数组无法与断言的字节值匹配。

File: base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java [hunk 7]
```diff
@@ -594,9 +594,9 @@ public void test_encodeObjectToBase64_WithJavaBean() throws Exception {
         user.setEmail("1484036491@qq.com");
         String base64 = Utils.encodeObjectToBase64(user);
         assertNotNull(base64);
-        assertFalse(base64.contains("lulu"));
-        assertFalse(base64.contains("22"));
-        assertFalse(base64.contains("image/"));
+        String decoded = new String(Base64.getDecoder().decode(base64), StandardCharsets.UTF_8);
+        assertTrue(decoded.contains("msslulu"));
+        assertTrue(decoded.contains("1484036491@qq.com"));
```
Comment (chilingling) [inline base/src/test/java/com/tinyengine/it/common/utils/UtilsTest.java]: 不要写入自己的个人信息

**Review by chilingling — CHANGES_REQUESTED** [review]

**Review by hexqi — APPROVED** [review]

**Review by chilingling — APPROVED** [review]
